Fix cleanup of old flow cells

islean opened the pull request:
## Description

Addresses issue #2170. A try/except block was added to ensure that the cleanup is not aborted when removal of one old flow cell directory fails. 

### Fixed

- Changed how we find already archived sample sheets, since there were inconsistencies with the tagging. Now we find sample sheets which are tagged with "samplesheet" as well as "archived_sample_sheet".

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b fix-clean-old-flow-cells -a
    ```

### How to test

- [x] Copy flow cells from production to stage which should be deleted.
- [x] Change tagging of one of the flow cells so that it won't be noticed by the filter.
- [x] Run a clean up of old flow cells in stage.

### Expected test outcome

- [ ] Check that IntegrityError is caught.
- [ ] Check that the remaining flow cells still get archived and removed.
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

islean commented:
Also - the IntegrityError will be different depending on if it is run on hasta or not. Locally it will be an sqlite IntegrityError excepted while on Hasta it should be SQLAlchemy.

seallard reviewed (approved):
Nice 👍

islean commented:
Currently the test written 'actually' deletes the flow cell, meaning that you need to rollback after every test, so this needs to be fixed before merging.

diitaz93 commented on cg/meta/clean/flow_cell_run_directories.py:
```diff
@@ -104,15 +103,42 @@ def archive_sample_sheet(self) -> None:
             tags=hk_tags,
         )
+
+    def sample_sheets_from_latest_version(self, hk_bundle_name: str) -> List[File]:
```
function name should start with a verb
```suggestion
    def get_sample_sheets_from_latest_version(self, hk_bundle_name: str) -> List[File]:
```

diitaz93 commented on cg/meta/clean/flow_cell_run_directories.py:
```diff
@@ -104,15 +103,42 @@ def archive_sample_sheet(self) -> None:
         LOG.info("Sample sheet found!")
         hk_bundle: Bundle = self.hk.bundle(name=self.id)
-        hk_tags: List[str] = [SequencingFileTag.ARCHIVED_SAMPLE_SHEET, self.id]
+        hk_tags: List[str] = [
+            self.id,
+        ]
```
This can be set up into one line. Probably Black did that formatting before you modified the line

diitaz93 reviewed (commented):
Looks good, I would change the implementation of the function `filter_on_sample_sheets`

islean commented on cg/meta/clean/flow_cell_run_directories.py:
```diff
@@ -104,15 +103,42 @@ def archive_sample_sheet(self) -> None:
+
+
+def filter_on_sample_sheets(list_of_files: List[File]):
+    return_files = list_of_files
```
No, that is probably more logical. My intuition said "filter should remove" and my brain stopped working there.

diitaz93 reviewed (approved):
looks good again! 👍

islean commented on tests/meta/clean/test_remove_old_dirs.py:
```diff
@@ -0,0 +1,78 @@
+):
+    # GIVEN an old flow cell which is not tagged as archived but whose sample sheet is included in Housekeeper
+
+    bundle_data: Dict = {"name": "HLG5GDRXY", "created": timestamp, "files": []}
+
+    helpers.ensure_hk_bundle(store=real_housekeeper_api, bundle_data=bundle_data)
+
+    real_housekeeper_api.add_and_include_file_to_latest_version(
+        bundle_name="HLG5GDRXY",
+        file=Path(dragen_flow_cell_dir, "SampleSheet.csv"),
+        tags=["samplesheet", "HLG5GDRXY"],
+    )
```
What would be the benefit of having it as a fixture? Won't it be maintained for the whole test suite then since it is not used anywhere else that I know of?

islean commented:
Test worked as hoped for:

<img width="1440" alt="Screenshot 2023-06-29 at 16 01 26" src="https://github.com/Clinical-Genomics/cg/assets/40887124/b56a2dd5-c83a-49db-b660-56185caa5bab">

A flow cell whose sample sheet was only tagged with "samplesheet" and not "archived_sample_sheet" was still found in hk and the run directory was removed.

islean commented on cg/meta/clean/flow_cell_run_directories.py:
```diff
@@ -104,15 +103,41 @@ def archive_sample_sheet(self) -> None:
+        if sample_sheets_from_latest_version:
+            for file in sample_sheets_from_latest_version:
+                if file.is_included:
+                    LOG.warning("Sample sheet already included!")
```
I agree in principle but in this case, the log just prior to is "LOG.info(f"Start archiving sample sheet {self.sample_sheet_path}")" so I don't think there is a reason to include it here as well. You can see how it looks in the screen shot from the test.

islean commented on cg/meta/clean/flow_cell_run_directories.py:
```diff
@@ -104,15 +103,41 @@ def archive_sample_sheet(self) -> None:
             tags=hk_tags,
         )
+
+    def get_sample_sheets_from_latest_version(self, hk_bundle_name: str) -> List[File]:
+        """Returns the files tagged with samplesheet or archived_sample_sheet for the given bundle."""
+        files: List[File] = self.hk.get_files_from_latest_version(
+            bundle_name=hk_bundle_name, tags=[self.id]
+        ).all()
+        return self.filter_on_sample_sheets(files=files)
+
+    @staticmethod
+    def filter_on_sample_sheets(files: List[File]) -> List[File]:
+        """Returns a list of the files given which are tagged with samplesheet or archived_sample_sheet."""
+        files_with_a_sample_sheet_tag: List[File] = []
+        for file in files:
+            file_tag_names: List[str] = [tag.name for tag in file.tags]
+            if (
+                SequencingFileTag.SAMPLE_SHEET in file_tag_names
+                or SequencingFileTag.ARCHIVED_SAMPLE_SHEET in file_tag_names
+            ):
+                files_with_a_sample_sheet_tag.append(file)
+        return [*set(files_with_a_sample_sheet_tag)]
```
Yeah, there is a function for filtering on "all" tags specified. I don't really know what is preferable here - feels a bit niche to have a function for querying with "mandatory" tags and "must have at least one of the following" tags, although that would potentially decrease the time and space complexity.

Vince-janv reviewed (approved):
Clean fix, and nice tests!

islean commented on cg/meta/clean/flow_cell_run_directories.py:
```diff
@@ -104,15 +103,41 @@ def archive_sample_sheet(self) -> None:
+        sample_sheets_from_latest_version: Optional[
+            List[File]
+        ] = self.get_sample_sheets_from_latest_version(hk_bundle_name=hk_bundle.name)
+        if sample_sheets_from_latest_version:
```
Yeah, just saw that when I was switching to "samplesheet" instead of "archived_sample_sheet" since that's what @karlnyr suggested.

islean replied:
Good catch! Not necessary at all but I got confused by the erroneous type hint. get_sample_sheets_from_latest_version will at worst return an empty list, not None. If it would return None, the if statement would be necessary.

islean commented on cg/meta/clean/flow_cell_run_directories.py:
```diff
@@ -104,15 +103,41 @@ def archive_sample_sheet(self) -> None:
         LOG.info("Sample sheet found!")
         hk_bundle: Bundle = self.hk.bundle(name=self.id)
-        hk_tags: List[str] = [SequencingFileTag.ARCHIVED_SAMPLE_SHEET, self.id]
+        hk_tags: List[str] = [self.id]
```
Yeah, saw that when I swapped to SAMPLESHEET instead of ARCHIVED_SAMPLE_SHEET, switched that now. Old artifact...

islean commented:
Performed another test now, similar to the last. Since I changed how tags are detected, I wanted to verify that it still adds samplesheets to HK when no files tagged with "samplesheet" or "archived_sample_sheet" were in the latest version. This worked (although it broke due to me reusing a flow cell which was already included but I removed a tag.):

```
] 3m58s cg clean remove-old-flow-cell-run-dirs -o 2000 -s hiseqx
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Checking directory /home/proj/stage/flowcells/hiseqx of sequencer hiseqx:
Checking flow cell 180522_ST-E00198_0301_BHLCKNCCXY
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
Found flow cell HLCKNCCXY in Statusdb, getting sequenced date.
Flow cell /home/proj/stage/flowcells/hiseqx/180522_ST-E00198_0301_BHLCKNCCXY is 1865 days old and will NOT be removed.
Checking flow cell 170407_ST-E00198_0209_BHHKVCALXX
Found flow cell HHKVCALXX in Statusdb, getting sequenced date.
Flow cell /home/proj/stage/flowcells/hiseqx/170407_ST-E00198_0209_BHHKVCALXX is 2272 days old and will be removed.
Removing flow cell run directory /home/proj/stage/flowcells/hiseqx/170407_ST-E00198_0209_BHHKVCALXX.
Start archiving sample sheet /home/proj/stage/flowcells/hiseqx/170407_ST-E00198_0209_BHHKVCALXX/SampleSheet.csv
Sample sheet found!
Fetching bundle with name: HHKVCALXX
Fetching files with tags in [HHKVCALXX]
Fetching files from version 130872
Fetching tag with name: HHKVCALXX
Fetching tag with name: samplesheet
Fetching tag with name: HHKVCALXX
Fetching tag with name: samplesheet
Created new bundle version dir: /home/proj/stage/housekeeper-bundles/HHKVCALXX/2023-02-09
Another file with identical included file path: /home/proj/stage/housekeeper-bundles/HHKVCALXX/2023-02-09/SampleSheet.csv already exist. Skip linking of: /home/proj/stage/flowcells/hiseqx/170407_ST-E00198_0209_BHHKVCALXX/SampleSheet.csv
...
sqlalchemy.exc.IntegrityError: (pymysql.err.IntegrityError) (1062, "Duplicate entry 'HHKVCALXX/2023-02-09/SampleSheet.csv' for key 'file.path'")
[SQL: INSERT INTO file (path, checksum, to_archive, version_id) VALUES (%(path)s, %(checksum)s, %(to_archive)s, %(version_id)s)]
[parameters: {'path': 'HHKVCALXX/2023-02-09/SampleSheet.csv', 'checksum': None, 'to_archive': 0, 'version_id': 130872}]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
```

islean commented:
Deployed to production by IO:

```
Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Updating 57bb8d6..8180a3c
Fast-forward
 .bumpversion.cfg                                         |  2 +-
 cg/__init__.py                                           |  2 +-
 cg/apps/housekeeper/hk.py                                |  4 +---
 cg/cli/clean.py                                          | 24 ++++++++++++------------
 cg/meta/clean/flow_cell_run_directories.py               | 43 +++++++++++++++++++++++++++++++++----------
 setup.py                                                 |  2 +-
 tests/meta/clean/test_clean_flow_cell_run_directories.py | 25 +------------------------
 tests/meta/clean/test_remove_old_dirs.py                 | 78 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 8 files changed, 128 insertions(+), 52 deletions(-)
 create mode 100644 tests/meta/clean/test_remove_old_dirs.py
/home/hiseq.clinical
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/hiseq.clinical
done.
Log deploy... done.
cg, version 37.0.27
θ72° 17d [hiseq.clinical@hasta:~] [P_base] 2m16s $ 
```
